lib/ansible/plugins/test/uri.py

```diff
@@ -0,0 +1,33 @@
+# (c) Ansible Project
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from urllib.parse import urlparse
+
+
+def is_uri(value, schemas=None):
+    ''' Will verify that the string passed is a valid 'uri' according to spec, if passed a list of valid schemas it will even match those '''
+    try:
+        x = urlparse(value)
+        isit = all([x.scheme, x.netloc, not schemas or x.scheme in schemas])
+    except Exception as e:
+        isit = False
+    return isit
+
+
+def is_url(value):
+    ''' Will verify that the string passed is a valid 'url' according to spec '''
+    return is_uri(value, ['http', 'ftp', 'https', 'ftps'])
```

Comment:
Please link to the spec you mean here. The spec I know of, https://url.spec.whatwg.org/, already mentions several more valid schemes (see for example https://url.spec.whatwg.org/#url-miscellaneous), and if I understand it correctly basically any non-empty scheme string is allowed.

Reply:
I mixed 'special scheme' with valid (also did not know websockets had been added to list, TIL), but mostly these are things we can deal with as urls as we have builtin handlers ... might need to give more options to user to handle 'unusable vs usable' in the context of urllib

Reply:
I was going for mostly 'commonly usefule' but for full list of supported either using the urllib implementaiton
https://github.com/python/cpython/blob/3.10/Lib/urllib/parse.py#L47
or 'official list' https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml
but i'll just 'opt out' and leave it up to user